src/test/java/io/pravega/connectors/flink/FlinkPravegaTableSourceTest.java

```diff
@@ -171,6 +174,51 @@ public void testTableSourceDescriptor() {
         TableSourceUtil.validateTableSource(actualSource);
     }
 
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testTableSourceDescriptorWithWatermark() {
+        final String cityName = "fruitName";
+        final String total = "count";
+        final String eventTime = "eventTime";
+        final String procTime = "procTime";
+        final String controllerUri = "tcp://localhost:9090";
+        final long delay = 3000L;
+        final String streamName = "test";
+        final String scopeName = "test";
+
+        Stream stream = Stream.of(scopeName, streamName);
+        PravegaConfig pravegaConfig = PravegaConfig.fromDefaults()
+                .withControllerURI(URI.create(controllerUri))
+                .withDefaultScope(scopeName);
+
+        // construct table source using descriptors and table source factory
+        Pravega pravega = new Pravega();
+        pravega.tableSourceReaderBuilder()
+                .withTimestampAndWatermark(new MyAssigner())
```

Comment:
I have a concern about the table API, could you provide some links about how the Kafka table source does it?   One specific concern is whether the `AssignerWithTimeWindows` is an appropriate interface for this situation.  I recall there's `rowtime` stuff which tells the table API about which column contains the event time.

Reply:
Yes, Flink has a `rowtime` which tells the timestamp and watermark. However, this only support watermark of `periodicTimestampAssigner` or `punctuatedTimestampAssigner` as its strategy, See: https://github.com/apache/flink/blob/master/flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/wmstrategies/WatermarkStrategy.java#L33. Both of them are data-driven watermark and it is totally different from our Pravega's watermark. It is a good news that it supports underlying `dataStreamSource` to have its own timestamp and watermark, so I choose to implement it in the underlying stream. User can use ````

Reply:
To answer your first question, Kafka did nothing to support its own watermark strategy in Table API. It only put effort in their normal consumer at https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java#L293 . I can find nothing related to timestamp or watermark in its table API's properties, I think Flink treats Table API's row time attributes as data-driven mostly.